File: frontend/src/api/modules/key-usage.ts
import type { RequestOptions } from '../request'
import type { DashboardHealthTimeline } from './dashboard'
import type { UsageBilling, UsageLatencyDetails, UsageTokenDetails } from './usage'
import request from '../request'

export interface KeyUsageConfig {
  name: string
  plaintextKey: string
}

export interface KeyUsageVersion {
  version: string
  gitSha: string
}

export interface KeyUsageMetrics {
  requests: number
  inputTokens: number
  outputTokens: number
  cachedTokens: number
  cacheWriteTokens: number
  reasoningTokens: number
  totalTokens: number
  costUsd: string | null
  costIncomplete: boolean
}

export interface KeyUsageBudget {
  seatName: string | null
  accountCycle: boolean
  name: string
  prefix: string
  maxConcurrency: number
  requestsPerMinute: number
  dailyLimitUsd: string
  dailyUsedUsd: string
  dailyResetsAt: string | null
  weeklyLimitUsd: string
  weeklyUsedUsd: string
  weeklyResetsAt: string | null
}

export interface SeatKeyUsage {
  id: string
  name: string
  prefix: string
  current: boolean
  revoked: boolean
  dailyUsedUsd: string
  weeklyUsedUsd: string
}

export interface KeyUsageTrendPoint extends KeyUsageMetrics {
  time: string
  bucketSeconds: number
}

export interface KeyUsageOverview {
  asOf: string
  startTime: string
  endTime: string
  key: KeyUsageBudget
  seatKeys: SeatKeyUsage[]
  summary: KeyUsageMetrics
  models: KeyUsageModel[]
  modelsPagination: { currentPage: number, pageSize: number, hasMore: boolean }
  trend: KeyUsageTrendPoint[]
  healthTimeline: DashboardHealthTimeline
}

export interface KeyUsageModel {
  model: string
  requests: number
  totalTokens: number
  costUsd: string | null
  costIncomplete: boolean
}

export type KeyUsageRecordKind = 'success' | 'error'

export interface KeyUsageRecord {
  id: string
  createdAt: string
  model: string | null
  route: string | null
  reasoningEffort: string | null
  clientTransport: string | null
  upstreamTransport: string | null
  tokenDetails: UsageTokenDetails | null
  billing: UsageBilling | null
  latencyMs: number | null
  firstTokenLatencyMs: number | null
  latencyDetails: Pick<UsageLatencyDetails, 'firstEventMs' | 'firstReasoningMs' | 'firstTextMs'>
  clientIp: string | null
  userAgent: string | null
  status: KeyUsageRecordKind
  statusCode: number | null
}

export interface KeyUsagePage {
  items: KeyUsageRecord[]
  currentPage: number
  pageSize: number
  total: number
}

export interface KeyUsageQuery {
  startTime: string
  endTime: string
  model?: string
  currentPage?: number
  pageSize?: number
}

export function getKeyUsageOverview(params: KeyUsageQuery, options: RequestOptions = {}) {
  return request<KeyUsageOverview>({
    url: '/api/key-usage/overview',
    method: 'GET',
    params,
    ...options,
  })
}

export function getKeyUsageVersion(options: RequestOptions = {}) {
  return request<KeyUsageVersion>({
    url: '/api/key-usage/version',
    method: 'GET',
    ...options,
  })
}

export function getKeyUsageConfig(options: RequestOptions = {}) {
  return request<KeyUsageConfig>({
    url: '/api/key-usage/config',
    method: 'GET',
    ...options,
  })
}

export function getKeyUsageRecords(
  params: KeyUsageQuery & { kind: KeyUsageRecordKind, currentPage: number, pageSize: number },
  options: RequestOptions = {},
) {
  return request<KeyUsagePage>({
    url: '/api/key-usage/records',
    method: 'GET',
    params,
    ...options,
  })
}

File: backend/crates/gateway-admin/src/model/key_usage.rs
//! 单页 Key 用量的查询合同；身份范围不由调用方提供。

use super::{
    PageSize,
    client_keys::ClientKeyRecord,
    observability::{
        DiagnosticObservationPage, DiagnosticPageQuery, HealthTimeline, OpsErrorPage,
        RequestMetricPoint, TimeRange, UsageOverview, UsagePage,
    },
};
use gateway_core::{metering::Decimal, policy::ClientApiKeyId};

#[derive(Debug, Clone)]
pub struct KeyUsageQuery {
    pub range: TimeRange,
    pub model: Option<String>,
}

#[derive(Debug, Clone)]
pub struct KeyUsageOverviewQuery {
    pub usage: KeyUsageQuery,
    pub models_page: DiagnosticPageQuery,
}

#[derive(Debug, Clone, Copy)]
pub enum KeyUsageRecordKind {
    Success,
    Error,
}

#[derive(Debug, Clone)]
pub struct KeyUsageRecordsQuery {
    pub usage: KeyUsageQuery,
    pub kind: KeyUsageRecordKind,
    pub current_page: u32,
    pub page_size: PageSize,
}

pub struct KeyUsageOverview {
    pub key: ClientKeyRecord,
    pub seat_keys: Vec<SeatKeyUsage>,
    pub overview: UsageOverview,
    pub trend: Vec<RequestMetricPoint>,
    pub models: DiagnosticObservationPage,
    pub health_timeline: HealthTimeline,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SeatKeyUsage {
    pub id: ClientApiKeyId,
    pub name: String,
    pub prefix: String,
    pub revoked: bool,
    pub daily_used_usd: Decimal,
    pub weekly_used_usd: Decimal,
}

pub enum KeyUsageRecords {
    Success(UsagePage),
    Error(OpsErrorPage),
}
